apple-ib-tb.c

```diff
@@ -319,6 +339,77 @@ static int appletb_send_hid_report(struct appletb_iface_info *iface_info,
 	return rc;
 }
 
+static int apple_magic_keyboard_backlight_set(struct apple_magic_backlight *backlight, char brightness, char rate)
+{
+	int tries = 0;
+	int rc;
+	struct apple_magic_keyboard_backlight_report *rep;
+
+	rep = kmalloc(sizeof(*rep), GFP_KERNEL);
+	if (rep == NULL)
+		return -ENOMEM;
+
+	rep->report_id = 0x01;
+	rep->mode = brightness;
+	rep->brightness = brightness;
+	rep->max = 0x5e;
+	rep->rate = rate;
+
+	do {
+		/*
+		 * FIXME: use appletb_send_hid_report, don't hard code all of this
+		 * Need to get apple_tb_send_hid_report to use wIndex=0x01
+		 */
+		rc = usb_control_msg(&backlight->dev,
+			usb_sndctrlpipe(&backlight->dev, 0),
+			HID_REQ_SET_REPORT, USB_DIR_OUT |
+			USB_TYPE_CLASS | USB_RECIP_INTERFACE,
+			0x0301, 0x01, rep, sizeof(*rep),
+			2000);
+		if (rc != -EPIPE)
+			break;
+
+		usleep_range(1000 << tries, 3000 << tries);
+	} while (++tries < 5);
+
+	kfree(rep);
+	return (rc > 0) ? 0 : rc;
+}
+
+static int apple_magic_keyboard_backlight_led_set(struct led_classdev *led_cdev,
+		enum led_brightness brightness)
+{
+	struct apple_magic_backlight *backlight = container_of(led_cdev,
+			struct apple_magic_backlight, cdev);
+
+	return apple_magic_keyboard_backlight_set(backlight, brightness, 1);
+}
+
+static int apple_magic_keyboard_backlight_init(struct appletb_device *tb_dev)
+{
+	int ret;
+	struct apple_magic_backlight *backlight;
+
+	backlight = devm_kzalloc(tb_dev->log_dev, sizeof(*backlight), GFP_KERNEL);
+	if (!backlight)
+		return -ENOMEM;
+
+	backlight->dev = *interface_to_usbdev(tb_dev->disp_iface.usb_iface);
+	backlight->cdev.name = "apple::kbd_backlight";
+	backlight->cdev.max_brightness = APPLE_MAGIC_KBD_BL_MAX;
+	backlight->cdev.brightness_set_blocking = apple_magic_keyboard_backlight_led_set;
+
+	ret = apple_magic_keyboard_backlight_set(backlight, 0, 0);
+	if (ret) {
+		dev_err(tb_dev->log_dev, "Failed to initialise Magic Keyboard Backlight (%d)\n", ret);
+		kfree(backlight);
+		return ret;
+	}
+
+	ret = devm_led_classdev_register(tb_dev->log_dev, &backlight->cdev);
+	return ret;
```

Comment:
Shouldn't backlight structure be freed here?

Reply:
I've put the backlight struct into tb_dev. The class dev seems to get unregistered for us (when I manually deregister it there's a null pointer)